src/ring_theory/subsemiring.lean

```diff
@@ -339,6 +390,44 @@ lemma closure_eq_of_le {s : set R} {t : subsemiring R} (h₁ : s ⊆ t) (h₂ :
   closure s = t :=
 le_antisymm (closure_le.2 h₁) h₂
 
+end subsemiring
+
+namespace submonoid
+
+lemma to_subsemiring_coe :
+  (M.to_subsemiring : set R) = add_submonoid.closure (M : set R) := rfl
+
+lemma to_subsemiring_to_add_submonoid :
+  M.to_subsemiring.to_add_submonoid = add_submonoid.closure (M : set R) := rfl
+
+/-- The `subsemiring` generated by a multiplicative submonoid coincides with the
+`subsemiring.closure` submonoid itself . -/
+lemma to_subsemiring_eq_closure : M.to_subsemiring = subsemiring.closure (M : set R) :=
+begin
+  ext,
+  refine ⟨λ hx, _, λ hx, (subsemiring.mem_closure.mp hx) M.to_subsemiring (λ s sM, _)⟩;
+  rintros - ⟨H1, rfl⟩;
+  rintros - ⟨H2, rfl⟩,
+  { exact add_submonoid.mem_closure.mp hx H1.to_add_submonoid H2 },
+  { exact H2 sM }
+end
+
+end submonoid
+
+namespace subsemiring
+
+@[simp]
+lemma closure_submonoid_closure {s : set R} : closure ↑(submonoid.closure s) = closure s :=
```

Comment:
The following is also true right? Might be nice to prove it for consistency
```lean
@[simp]
lemma closure_add_submonoid_closure {s : set R} : closure ↑(add_submonoid.closure s) = closure s :=
```

Reply:
Eric, I think that I proved what you wanted.  It may not be the nicest proof, though: feel free to suggest a better approach!

Reply:
I had a go at a neater proof, but couldn't come up with anything better. What I did do was move some of your lemmas to `submonoid/membership`.